tools/dds/dds-sniffer/dds-sniffer.cpp

```diff
@@ -144,159 +159,410 @@ void Sniffer::run( uint32_t seconds )
+        if( _print_discoveries )
+        {
+            print_reader_discovered( info, true );
+        }
+        save_topic_reader( info );
+        break;
+    case eprosima::fastrtps::rtps::ReaderDiscoveryInfo::REMOVED_READER:
+        if( _print_discoveries )
+        {
+            print_reader_discovered( info, false );
+        }
+        remove_topic_reader( info );
+        break;
+    }
+}
 
-        // Create a data reader for the topic
-        if( _subscriber == nullptr )
+void dds_sniffer::on_participant_discovery( eprosima::fastdds::dds::DomainParticipant *,
+                                            eprosima::fastrtps::rtps::ParticipantDiscoveryInfo && info )
+{
+    switch( info.status )
+    {
+    case eprosima::fastrtps::rtps::ParticipantDiscoveryInfo::DISCOVERED_PARTICIPANT:
+        if( _print_discoveries )
         {
-            _subscriber = _participant->create_subscriber( eprosima::fastdds::dds::SUBSCRIBER_QOS_DEFAULT, nullptr );
+            print_participant_discovered( info, true );
+        }
+        _discovered_participants[info.info.m_guid] = info.info.m_participantName;
+        break;
+    case eprosima::fastrtps::rtps::ParticipantDiscoveryInfo::REMOVED_PARTICIPANT:
+    case eprosima::fastrtps::rtps::ParticipantDiscoveryInfo::DROPPED_PARTICIPANT:
+        if( _print_discoveries )
+        {
+            print_participant_discovered( info, false );
+        }
+        _discovered_participants.erase( info.info.m_guid );
+        break;
+    }
+}
 
-            if( _subscriber == nullptr )
-            {
-                return;
-            }
+void dds_sniffer::on_type_information_received( eprosima::fastdds::dds::DomainParticipant *,
+                                                const eprosima::fastrtps::string_255 topic_name,
+                                                const eprosima::fastrtps::string_255 type_name,
+                                                const eprosima::fastrtps::types::TypeInformation & type_information )
+{
+    std::cout << "Topic '" << topic_name << "' with type '" << type_name << "' received" << std::endl;
+}
+
+void dds_sniffer::save_topic_writer( const eprosima::fastrtps::rtps::WriterDiscoveryInfo & info )
+{
+    std::string topic_name = info.info.topicName().c_str();
+    _discovered_topics[topic_name].writers.insert( info.info.guid() );
+    save_max_indentation( std::move( topic_name ) );
+}
+
+void dds_sniffer::remove_topic_writer( const eprosima::fastrtps::rtps::WriterDiscoveryInfo & info )
+{
+    auto topic_entry = _discovered_topics[info.info.topicName().c_str()];
+    auto writer_iter = topic_entry.writers.find( info.info.guid() );
+    if( writer_iter != topic_entry.writers.end() )
+    {
+        topic_entry.writers.erase( writer_iter );
+        if( topic_entry.writers.empty() && topic_entry.readers.empty() )
+        {
+            _discovered_topics.erase( info.info.topicName().c_str() );
         }
+    }
+}
```

Comment:
Don't you want to update the max indent when topics are removed? Really we need the indent only for snapshot, yes? Then why not calculate it once just before the dump, rather than having to worry about it one each writer/reader add/remove?

Reply:
Calculating max indentation only when printing